File: Medium/24_SwapNodesInPairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 * Basically, need a pointer point to current head and do the swap.
 * Then, move the pointer forward
 */
let swapPairs = function (head) {
    if (head === null || (head !== null && head.next === null)) return head;
    let current = head;
    let next = head.next;
    head.next = next.next;
    next.next = current;
    current = head;
    let result = next;
    while (current.next !== null && current.next.next !== null) {
        let nextHead = current.next;
        next = nextHead.next;
        nextHead.next = next.next;
        next.next = nextHead;
        current.next = next;
        current = nextHead;
    }
    return result;
};

File: Medium/24_swap-nodes-in-pairs.java
package Medium;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummyNode = new ListNode(-1);
        dummyNode.next = head;

        ListNode node = dummyNode;

        while (node.next != null && node.next.next != null) {
            ListNode prev = node;
            ListNode curNode = prev.next;

            ListNode temp = curNode.next;
            curNode.next = temp.next;
            temp.next = prev.next;
            prev.next = temp;

            node = curNode;
        }

        return dummyNode.next;
    }
}
